gobblin-core/src/main/java/gobblin/source/extractor/hadoop/CopySource.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.hadoop;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.SourceState;
+import gobblin.configuration.State;
+import gobblin.configuration.WorkUnitState;
+import gobblin.data.management.copy.CopyableDataset;
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.data.management.dataset.Dataset;
+import gobblin.data.management.dataset.DatasetUtils;
+import gobblin.data.management.partition.Partition;
+import gobblin.data.management.partition.PartitionableDataset;
+import gobblin.data.management.retention.dataset.finder.DatasetFinder;
+import gobblin.source.extractor.Extractor;
+import gobblin.source.extractor.extract.AbstractSource;
+import gobblin.source.workunit.Extract;
+import gobblin.source.workunit.WorkUnit;
+import gobblin.util.ForkOperatorUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Collection;
+import java.util.List;
+
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+
+/**
+ * {@link gobblin.source.Source} that generates work units from {@link gobblin.data.management.copy.CopyableDataset}s.
+ * Used for doing byte-level copies of files.
+ */
+public class CopySource extends AbstractSource<String, byte[]> {
```

Comment:
I don't think the record type should be byte[]. I think it should be a custom object that contains:
* An InputStream or FSDataInputStream of the input file. Chances are that the input stream we are reading from is already buffered so we are adding double-buffering overhead. Further, having the InputStream data model will allow us to do some transformation like encrypting/decrypting, compressing/decompressing on-the-fly. I think that FSDataInputStream might be a better fit since it implements Seekable. I think we need that if we want to duplicate input streams (for fork).
* OwnerAndPermission
* Other metadata (e.g. support for extended attributes if we move to a version of hadoop that supports it).